Rewrite some decomps to allow producing aten ops

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93099
* #93130

This introduces a new stop to the decomposition train.
Before reaching prims.view_of, it will stop at aten.alias. Export path wants to get off the train at aten ops.

SherlockNoMad committed:
Rewrite some decomps to allow producing aten ops

[ghstack-poisoned]

ngimel commented on torch/_prims_common/wrappers.py:
```diff
@@ -39,7 +39,7 @@ def _maybe_convert_to_dtype(a, dtype):
             # NOTE: this is incorrect on the CPU
             # See https://github.com/pytorch/pytorch/issues/77553
-            return prims.convert_element_type(a, dtype)
+            return a.to(dtype)
```
This could increase tracing time, as `to` would go through a series of redispatches, cc @chillee, @voznesenskym

Chillee replied:
What's the motivation behind this PR?

ngimel replied:
Backends that want only `aten` ops in the graph they are getting.

SherlockNoMad replied:
This introduces a new stop to the decomposition train. 
Before reaching prims.view_of, it will stop at aten.alias. Export path want to get off train at aten ops.

Chillee replied:
@SherlockNoMad imo, we should just turn `prims.convert_element_type` into an ATen op. Note that `aten.to.dtype` is  *not* a composite explicit op - it decomposes into `aten._to_copy`.

SherlockNoMad replied:
In this PR, I will directly use aten._to_copy instead of aten._to_copy. 
Agree that prims.convert_element_type should just be something like aten.cast.

ngimel replied:
This introduces `aten` op to `_refs` decompositions that are meant to operate on `torch.*` API. Conceptually `.to` was better, but it introduced yet another trip through dispatched ops. Probably land it with `.to` and let's see if it regresses tracing time.

SherlockNoMad replied:
OK, I reverted back to use .to()

SherlockNoMad committed:
Update on "Rewrite some decomps to allow producing aten ops"




[ghstack-poisoned]

SherlockNoMad committed:
Update on "Rewrite some decomps to allow producing aten ops"


This introduces a new stop to the decomposition train.
Before reaching prims.view_of, it will stop at aten.alias. Export path wants to get off the train at aten ops.





[ghstack-poisoned]

zhxchen17 commented:
just for reference and for record, we got some failures internally, diff number: D42764068

SherlockNoMad commented:
> just for reference and for record, we got some failures internally, diff number: D42764068

Noted, I will hold off landing https://github.com/pytorch/pytorch/pull/93133 until failures are fixed.

SherlockNoMad committed:
Update on "Rewrite some decomps to allow producing aten ops"


This introduces a new stop to the decomposition train.
Before reaching prims.view_of, it will stop at aten.alias. Export path wants to get off the train at aten ops.





[ghstack-poisoned]

SherlockNoMad committed:
Update on "Rewrite some decomps to allow producing aten ops"


This introduces a new stop to the decomposition train.
Before reaching prims.view_of, it will stop at aten.alias. Export path wants to get off the train at aten ops.





[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge

ngimel commented:
Stamped to unblock, but are you sure this will actually help them? `to` will go to `_to_copy`, and if the use `_to_copy` decomposition it'll end up at the same `prims.convert_element_dtype` op. Are they stopping decomp at `_to_copy`?

SherlockNoMad commented:
> Stamped to unblock, but are you sure this will actually help them? `to` will go to `_to_copy`, and if the use `_to_copy` decomposition it'll end up at the same `prims.convert_element_dtype` op. Are they stopping decomp at `_to_copy`? Changing element dtype (and just element dtype) is an op that conceptually should be prim, that's why @Chillee suggested adding it to ATen, and all backends should support it.

I have a following PR to stop decomposing at aten._to_copy. https://github.com/pytorch/pytorch/pull/93131

SherlockNoMad commented:
@pytorchbot merge -f "flaky tests"
